### PR TITLE
RideShare - Anna Elisabeth - Queues

### PR DESCRIPTION
**Pull request by annaewolf**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    Design decision: how much information should Driver and Rider have about trips? I tried to move away from giving Driver and Rider a lot of information about trips, but ended up creating a lot of dependency between the classes despite my best efforts. I did move the .driver_find and rider_find methods to trips, rather than giving Driver and Rider all the information on trips and that was my most conscious design decision. However, Driver and Rider know that trips must have a duration and a cost and a rating and call on those methods within their class, so it's a very tightly coupled project.   |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  the difference between tightly and loosely coupled! it was fairly conceptual as I was reading Metz, but we hadn't done a project that required us to connect the classes so much. If I break one of my methods within trips, driver and rider will **have** to change.   |
| Describe a _nominal test_ that you wrote for this assignment. |  if driver.all can return the first line in the csv  |
| Describe an _edge case test_ that you wrote for this assignment. |  if I could initialize a new driver with a string as an ID instead of name  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | fairly well! writing the tests was a lot easier after writing the pseudocode -- it let me gather my thoughts on what i was going to do, write tests that tested that thing, then write my code.  |

### CODE REVIEW

File: specs/drivers_spec.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
+require_relative 'spec_helper'
+
+describe "Driver class" do
+  let(:driver) { RideShare::Driver.new({name: "Ron Weasley", driver_id: 1})}
```
Comment (droberts-sea) [inline specs/drivers_spec.rb:4]: You've got enough test cases here that it starts to get hard to tell what's doing what. It might be worthwhile to add nested `describe` blocks here to break things up - perhaps one per method.

File: specs/drivers_spec.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  it " can return the last driver from the CSV" do
+    # skip
+    RideShare::Driver.find(100).id.must_equal 100
+  end
+
+  it "won't return a driver that doesn't exist" do
+    proc {
+      RideShare::Driver.find(1000).id
+    }.must_raise NoMethodError
```
Comment (droberts-sea) [inline specs/drivers_spec.rb:43]: It might be a little cleaner to say

```ruby
RideShare::Driver.find(1000).must_be_nil
```

File: specs/drivers_spec.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      proc {
+        driver.all_trips
+      }.must_raise ArgumentError
+  end
+
+  it "can find a single driver's average rating" do
+    ron = driver.average_rating
+    ron.must_be_kind_of Float
+    ron.must_equal 2.33
+  end
```
Comment (droberts-sea) [inline specs/drivers_spec.rb:62]: What does `average_raiting` do if the driver has no trips?

File: specs/trips_spec.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  it "won't return a trip that doesn't exist" do
+    proc {
+      RideShare::Trip.find(800000).id
+    }.must_raise NoMethodError
+  end
+
+  it "can find all trip instances for a given rider ID" do
+    # skip
+    RideShare::Trip.rider_find(1).length.must_equal 2
+  end
```
Comment (droberts-sea) [inline specs/trips_spec.rb:70]: What if there are no trips for a given rider?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   yes - good work!	|
|   Answer comprehension questions	|   yes	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline comments |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |  mostly - same comments as for `Driver` tests  |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | yes |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |    |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   yes |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   yes |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

Great work overall! Code is clean and readable, your tests cover most of the cases we were looking for, and your git habits look strong. Keep up the good work!
